modules/autotagging-commons/src/main/java/org/opensearch/rule/feature_value_resolver/CandidateFeatureValues.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rule.feature_value_resolver;
+
+import org.opensearch.rule.attribute_extractor.AttributeExtractor;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Represents candidate feature values for an attribute
+ */
+public class CandidateFeatureValues {
+
+    /**
+     * A list of sets of candidate feature values collected for an attribute
+     * The list is ordered from the most specific match to less specific ones. For example:
+     * featureValues = [ {"a", "b"}, {"c"} ]
+     * Here, {"a", "b"} comes first because these feature values comes from rules with a more specific match
+     * e.g. A rule with "username|123" is a more specific match than "username|1" when querying "username|1234".
+     */
+    private final List<Set<String>> featureValuesBySpecificity;
+
+    /**
+     * A flattened set of all candidate values collected across all specificity levels.
+     * This set combines all values in 'featureValues' into a single collection for easy access
+     * and intersection computations.
+     */
+    private final Set<String> flattenedValues = new HashSet<>();
+
+    /**
+     * Maps each feature value to the index of its first occurrence set in 'featureValues'.
+     * This helps in tie-breaking: values appearing earlier in the list (i.e., more specific matches)
+     * are considered better matches when resolving the final label.
+     */
+    private final Map<String, Integer> firstOccurrenceIndex = new HashMap<>();
```

Comment:
I am assuming this is for optimizing the lookup? Have we considered the latency impact without having this index?